### PR TITLE
Generate function body

### PR DESCRIPTION
**Pull request by EPashkin**
Almost full functions generation.
Last "trait constraint" and "return out parameters", but before I prefer get compilable gtk code.
If you don't start #21 then say, I try do it.

### CODE REVIEW

File: src/codegen/function_body.rs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        format!("{}ffi::{}({}){}", self.from_glib_prefix,
+            self.glib_name, param_str, self.from_glib_suffix)
+    }
+    fn generate_parameters(&self) -> String {
+        let mut param_str = String::with_capacity(500);
+
+        for (pos, par) in self.parameters.iter().enumerate() {
+            if pos > 0 { param_str.push_str(", ") }
+            param_str.push_str(&par);
+        }
```
Comment (gkoz) [inline src/codegen/function_body.rs]: Can this be done with `self.parameters.connect(", ")`?

File: src/codegen/function_body.rs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    from_glib_prefix: String,
+    from_glib_suffix: String,
+    parameters: Vec<String>,
+}
+
+impl Builder {
+    pub fn new() -> Builder {
+        Default::default()
+    }
+    pub fn glib_name(mut self, name: &str) -> Builder {
```
Comment (gkoz) [inline src/codegen/function_body.rs]: I don't think it matters here but FYI they [recommend against](https://aturon.github.io/ownership/builders.html#non-consuming-builders-(preferred):) consuming builders.

File: src/codegen/function_body.rs (hunk 1 above)
Comment (EPashkin) [inline src/codegen/function_body.rs]: Thank, don't know about it.

File: src/codegen/function.rs [hunk 3]
```diff
@@ -50,17 +52,17 @@ pub fn declaration(library: &library::Library, analysis: &analysis::functions::I
...
-    let mut v: Vec<String> = Vec::new();
-    //TODO: real generation
-    write_to_vec!(v, "unsafe {{");
-    write_to_vec!(v, "{}TODO: call ffi:{}()", tabs(1), analysis.glib_name);
-    write_to_vec!(v, "}}");
-    v
+    //TODO: change to map on parameters with pass Vec<String> to builder
+    for par in &analysis.parameters {
+        let s = par.to_parameter_body(library, in_trait);
+        builder = builder.parameter(s);
```
Comment (gkoz) [inline src/codegen/function.rs]: I guess this is some of the motivation for non-consuming builders: you'd just do `builder.parameter(s);` here.

File: src/codegen/function_body.rs (hunk 2 above)
Comment (EPashkin) [inline src/codegen/function_body.rs]: Thank again, forgot which is preferable

File: src/codegen/function.rs [hunk 4]
```diff
@@ -50,17 +52,17 @@ pub fn declaration(library: &library::Library, analysis: &analysis::functions::I
...
-macro_rules! write_to_vec(
-    ($dst:expr, $($arg:tt)*) => (
-        $dst.push(fmt::format(format_args!($($arg)*)))
-    )
-);
+pub fn body(library: &library::Library, analysis: &analysis::functions::Info,
+    in_trait: bool) -> Vec<String> {
+    let mut builder = Builder::new()
+        .glib_name(&analysis.glib_name)
+        .from_glib(analysis.ret.to_return_value_body(&library, &analysis));
```
Comment (gkoz) [inline src/codegen/function.rs]: Could you give a bit more thought to the names here? Wouldn't the `from_glib` and `to_return_value_body` methods have related names? Right now the only way to make sense of what happens here is to read a few function bodies.

Comment (EPashkin) [inline src/codegen/function.rs]: I do not like just 'from_glib' instead `to_return_value_body` for trait.
But for Builder It really means "conversion from glib".
Nothing better not come to mind.

**Comment by EPashkin** [comment]
Fixed builder and parameters joining.
Not changed `from_glib` and `to_return_value_body`

**Comment by gkoz** [comment]
Could `to_return_value_body` be though of as `translate_from_glib` or `to_glue_from_glib` because it generates the translation glue? Conversely the input parameters would be handled by `translate_to_glib`.

**Comment by EPashkin** [comment]
Only one "problem" translate_from_glib return "function" string and "translate_to_glib" - "method", 
may be needed some suffix?

File: src/codegen/function_body.rs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    from_glib_prefix: String,
+    from_glib_suffix: String,
+    parameters: Vec<String>,
+}
+
+impl Builder {
+    pub fn new() -> Builder {
+        Default::default()
+    }
+    pub fn glib_name<'a>(&'a mut self, name: &str) -> &'a mut Builder {
```
Comment (gkoz) [inline src/codegen/function_body.rs]: Did lifetime elision not handle this correctly? Reading the [TRPL](https://doc.rust-lang.org/book/lifetimes.html#lifetime-elision), it seems the return value would get the same lifetime as `self`.

**Comment by gkoz** [comment]
> Only one "problem" `translate_from_glib` return "function" string and `translate_to_glib` - "method",
may be needed some suffix?

Generally, it's possible that `translate_to_glib` may need to support the "function" syntax as well, although I can't think of an example currently.

**Comment by EPashkin** [comment]
Then it is better to make a difference now. `translate_from_glib_as_function` for method (don't changing trait name)?

**Comment by gkoz** [comment]
The method name seems okay. Not sure if there's point to having the trait name be different, shorter maybe.

**Comment by EPashkin** [comment]
Renamed. What about #21?

**Comment by EPashkin** [comment]
IMHO ready to merge

**Comment by gkoz** [comment]
Thanks.

**Pull request merged**